archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterMembersTest.java

```diff
@@ -0,0 +1,416 @@
+
+        assertThat(classes.get(SomeEnum.class).getMethodsWithReturnTypeOfSelf())
+                .extracting("name").contains("methodWithSomeEnumReturnType", "otherMethodWithSomeEnumReturnType");
+    }
+
+    @Test
+    public void classes_know_which_method_throws_clauses_contain_their_type() {
+        JavaClasses classes = new ClassFileImporter().importClasses(ClassWithThrowingMethod.class, FirstCheckedException.class);
+
+        Set<ThrowsDeclaration<JavaMethod>> throwsDeclarations = classes.get(FirstCheckedException.class).getMethodThrowsDeclarationsWithTypeOfSelf();
+        assertThatType(getOnlyElement(throwsDeclarations).getDeclaringClass()).matches(ClassWithThrowingMethod.class);
+        assertThat(classes.get(FirstCheckedException.class).getConstructorsWithParameterTypeOfSelf()).isEmpty();
+    }
+
+    @Test
+    public void classes_know_which_constructors_have_their_type_as_parameter() {
+        JavaClasses classes = new ClassFileImporter().importClasses(SomeClass.class, OtherClass.class, SomeEnum.class);
+
+        assertThat(classes.get(SomeEnum.class).getConstructorsWithParameterTypeOfSelf())
+                .extracting("owner").extracting("name")
+                .contains(SomeClass.class.getName(), OtherClass.class.getName());
+    }
+
+    @Test
+    public void classes_know_which_constructor_throws_clauses_contain_their_type() {
+        JavaClasses classes = new ClassFileImporter().importClasses(ClassWithThrowingConstructor.class, FirstCheckedException.class);
+
+        Set<ThrowsDeclaration<JavaConstructor>> throwsDeclarations =
+                classes.get(FirstCheckedException.class).getConstructorsWithThrowsDeclarationTypeOfSelf();
+        assertThatType(getOnlyElement(throwsDeclarations).getDeclaringClass()).matches(ClassWithThrowingConstructor.class);
+        assertThat(classes.get(FirstCheckedException.class).getMethodThrowsDeclarationsWithTypeOfSelf()).isEmpty();
+    }
+
+    @Test
+    public void classes_know_which_instanceof_checks_check_their_type() {
+        JavaClass clazz = new ClassFileImporter().importPackagesOf(InstanceofChecked.class).get(InstanceofChecked.class);
+
+        Set<JavaClass> origins = new HashSet<>();
+        for (InstanceofCheck instanceofCheck : clazz.getInstanceofChecksWithTypeOfSelf()) {
+            origins.add(instanceofCheck.getOwner().getOwner());
+        }
+        assertThatTypes(origins).matchInAnyOrder(ChecksInstanceofInMethod.class, ChecksInstanceofInConstructor.class, ChecksInstanceofInStaticInitializer.class);
+    }
+
+    @Test
+    public void automatically_resolves_field_types() {
+        @SuppressWarnings("unused")
+        class FieldTypeWithoutAnyFurtherReference {
+            String field;
+        }
+
+        JavaClass javaClass = new ClassFileImporter().importClass(FieldTypeWithoutAnyFurtherReference.class);
+
+        assertThat(javaClass.getField("field").getRawType().isFullyImported())
+                .as("field type is fully imported").isTrue();
```

Comment:
I like the new assertion, thanks a lot :smile: Only that in the second commit I would also remove the assertion for the modifiers then, because `matches` checks those as well :man_shrugging:

Reply:
Actually I'm gonna squash a little bit on top to extend `matches` for all these boolean flags... So then we also don't need to always check `isMemberClass()`? `isInnerClass()`? and so on. Because that should also simply match the object from the Reflection API.

Reply:
On second thought, maybe not :joy: Because those assertions actually depend on the context, then `matches(..)` can only be used on classes that have been directly imported, cause otherwise maybe the enclosing class is missing and thus `toplevelClass` or `memberClass` changes :disappointed: